tracee-ebpf/tracee/containers.go

```diff
@@ -0,0 +1,205 @@
+	if !ok {
+		c.mapContIds[contId] = make([]string, 0)
+	}
+}
+
+func (c *Containers) delContainer(contId string) {
+	delete(c.mapContIds, contId) // no-op if no contId
+}
+
+func (c *Containers) checkContainer(contId string) bool {
+	_, ok := c.mapContIds[contId]
+	return ok
+}
+
+func (c *Containers) GetContainers() []string {
+	var conts []string
+	for k := range c.mapContIds {
+		conts = append(conts, k)
+	}
+	if len(conts) == 0 {
+		return nil
+	}
+
+	return conts
+}
+
+func (c *Containers) addPid(contId string, pid string) {
+	c.addContainer(contId)
+	c.mapContIds[contId] = append(c.mapContIds[contId], pid)
+}
+
+func (c *Containers) GetPids(contId string) []string {
+	if !c.checkContainer(contId) {
+		return nil
+	}
+	return c.mapContIds[contId]
+}
+
+// Populate info
+
+func (c *Containers) Populate() error {
+	// do all the hard work
+
+	c.init()
+
+	err := c.procMountsCgroups()
+	if err != nil {
+		return err
+	}
+
+	return c.populate()
+}
+
+func (c *Containers) procMountsCgroups() error {
+	// find cgroups v1 and v2 mountpoints for procfs walks
+
+	mountsFile := "/proc/mounts"
+	file, err := os.Open(mountsFile)
+	if err != nil {
+		return err
+	}
+	scanner := bufio.NewScanner(file)
+	for i := 1; scanner.Scan(); i++ {
+		sline := strings.Split(scanner.Text(), " ")
+		mountpoint := sline[1]
+		fstype := sline[2]
+		if fstype == "cgroup" {
+			if strings.Contains(mountpoint, "cgroup/pids") {
+				c.cgroupv1 = true
+				c.cgroupv1mp = mountpoint
+			}
+		} else if fstype == "cgroup2" {
+			c.cgroupv2 = true
+			c.cgroupv2mp = mountpoint
+		}
+	}
+	_ = file.Close()
+
+	return nil
+}
+
+func (c *Containers) populate() error {
+	// Prepare regex to be used for population done in proc walk
+
+	if c.cgroupv1 {
+		var r []string
+		// docker:  <cgroupv1mp>/<random>/docker/<id>/tasks
+		r = append(r, ".*docker.*(.{64})/tasks$")
+		// podman:  <cgroupv1mp>/<random>/libpod-<id>.scope/tasks
+		r = append(r, ".*libpod.*(.{64})\\.scope/tasks$")
+		// generic: <cgroupv1mp>/<random>/<id>.scope/tasks
+		r = append(r, ".*(.{64})\\.scope/tasks$")
+
+		err := c.populateProcWalk(false, r)
+		if err != nil {
+			return err
+		}
+	}
+	if c.cgroupv2 {
+		var r []string
+		// docker:  <cgroupv2mp>/<random>/docker-<id>.scope/cgroup.procs
+		r = append(r, ".*docker.*(.{64})\\.scope/cgroup.procs$")
+		// podman:  <cgroupv2mp>/<random>/libpod-<id>.scope/cgroup.procs
+		r = append(r, ".*libpod.*(.{64})\\.scope/cgroup.procs$")
+		// generic: <cgroupv2mp>/<random>/<id>.scope/cgroup.procs
+		r = append(r, ".*(.{64})\\.scope/cgroup.procs$")
+
+		err := c.populateProcWalk(true, r)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Containers) populateProcWalk(v2 bool, allstr []string) error {
+	// Walk through procfs for cgroups v1 or v2 filesystem and find files based on regex
+	// Extract container information from found files
+
+	var basedir string
+	if v2 {
+		basedir = c.cgroupv2mp
+	} else {
+		basedir = c.cgroupv1mp
+	}
```

Comment:
You can give basedir as an argument instead of v2 (I don't see any other usage for v2 other than setting basedir)